Add NVM with Node 6 / 8 / 10, make Composer / Yarn / NPM builds optional

Cheddam opened the pull request:
This PR expands the image's flexibility, by adding [NVM](https://github.com/creationix/nvm) with Node v6, v8 and v10 preloaded. This enables the use of a `.nvmrc` file by projects to determine which version of Node should be used to compile their assets.

It also makes the build script determine whether to use Yarn or NPM, based on the existence of a `yarn.lock` file.

It _also_ makes the NPM / Yarn section of the build process entirely optional, based on the presence of `package.json`.

indygriffiths commented on Dockerfile:
```diff
@@ -13,6 +13,11 @@ RUN php -r "copy('https://getcomposer.org/installer', '/tmp/composer-setup.php')
 
+# NVM
+ENV NVM_DIR=/root/.nvm
+RUN curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.2/install.sh | bash
```
Personally a bit eh about piping a random Gist into bash

Cheddam replied:
That's a reasonable concern - but there aren't any good alternatives to this approach*, and it's no more dangerous than cloning the NVM repo and running the script from that, as both are trusting the developer of NVM to have appropriate security practices in place. Might need some feedback from @madmatt.

* - I _guess_ we could say "we trust this exact copy" and host it ourselves somewhere..

halkyon replied:
@madmatt Any thoughts here?

halkyon replied:
How about a checksum match on the file before executing it? I take it the file will never change because it's a tagged version.

indygriffiths commented:
@Cheddam Bit of team discussion that I want to run by you. Rather than introducing a new image and introducing Dash functionality, what if support for scripts/building assets is made in the default `platform-build` image and only run if the user has opted in?

In our mind this means:
- there is no need for Dash changes
- easier opt-in path for devs
- one image to manage instead of two

We'd prefer if the image _didn't_ automatically build static assets based on the presence of a package.json/lock file, as we don't want to start building assets for existing sites on deployment just in case it breaks something with their specific workflow.  Your idea of `predicate the whole NPM build step on whether a scp-build command is present in package.json` from the PR channel seems like the best solution to work out whether to build assets or not.

Cheddam commented:
@indygriffiths Forgot to push last week! I've added another dependency (which was available via APT, so should be a bit more trustworthy than NVM) called `jq` for querying the contents of the `package.json` file, and added a check for the presence of the `scp-build` command.

indygriffiths commented:
@Cheddam Do you want to close this PR off and make the changes on the platform-build repo instead if we're going to keep it as one image?

indygriffiths commented:
@Cheddam just checking in on this, all good if it's more of a wait-until-Hackday type of thing.

Cheddam commented:
Hey @indygriffiths, sorry for the lack of comms - my intent is to wait until we've discussed the final approach at the next Product Architecture Steering meeting before making any further changes.

Cheddam commented:
Superseded by a PR merged to silverstripe/platform-build.
